bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/Runtime.java

```diff
@@ -54,23 +61,46 @@ public static Runtime getCurrentRuntime() {
     }
 
 
-     /**
-      * Invoke Object method asynchronously. This will schedule the function and block the strand.
-      *
-      * @param object     Object Value.
-      * @param methodName Name of the method.
-      * @param strandName Name for newly creating strand which is used to execute the function pointer. This is optional
-      *                   and can be null.
-      * @param metadata   Meta data of new strand.
-      * @param callback   Callback which will get notify once method execution done.
-      * @param args       Ballerina function arguments.
-      * @return the result of the function invocation
-      */
-     public Object invokeMethodAsync(BObject object, String methodName, String strandName, StrandMetadata metadata,
-                                     Callback callback, Object... args) {
-         Function<?, ?> func = o -> object.call((Strand) (((Object[]) o)[0]), methodName, args);
-         return scheduler.schedule(new Object[1], func, null, callback, strandName, metadata).result;
-     }
+    /**
+     * Invoke Object method asynchronously. This will schedule the function and block the strand.
+     *
+     * @param object     Object Value.
+     * @param methodName Name of the method.
+     * @param strandName Name for newly creating strand which is used to execute the function pointer. This is optional
+     *                   and can be null.
+     * @param metadata   Meta data of new strand.
+     * @param callback   Callback which will get notify once method execution done.
+     * @param args       Ballerina function arguments.
+     * @return the result of the function invocation
+     */
+    public Object invokeMethodAsync(BObject object, String methodName, String strandName, StrandMetadata metadata,
+                                    Callback callback, Object... args) {
+        Function<?, ?> func = o -> object.call((Strand) (((Object[]) o)[0]), methodName, args);
+        boolean isIsolate = isIsolated(object.getType(), methodName);
+        if (isIsolate) {
+            return scheduler.schedule(new Object[1], func, null, callback, strandName, metadata).result;
+        } else {
+            return scheduler.scheduleToObjectGroup(object, new Object[1], func, null, callback, strandName, metadata)
+                    .result;
+        }
+    }
+
+    private boolean isIsolated(ObjectType type, String methodName) {
+        for (MethodType method : type.getMethods()) {
```

Comment:
Shall we keep the set of isolated object method/resource method names as a set inside ObjectType. We can get that list during compile time and set it during codegen. https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeGen.java#L1154

Reply:
The complete service object has to be isolated. This should come from the compiler frontend. 

Something similar to `ObjectType.isolated()` or a flag in the `ObjectType`. Checking all resource methods is not enough

Reply:
Modified to check both object and resource